# ESERCIZIO 6 e 7: JOIN 03.07
***

+ **Importare le librerie**

In [2]:
import numpy as np
import pandas as pd

+ **Scaricare dal sito ISTAT la popolazione residente al 1° gennaio e visualizzarla**

In [8]:
istat = pd.read_csv('../data/popolazioneISTAT-regioni1gen2020.csv', sep=';')
istat

Territorio   Maschi  Femmine    Totale
0                 Abruzzo   637388   668382   1305770
1              Basilicata   273941   282993    556934
2                Calabria   942644   982057   1924701
3                Campania  2820555  2965306   5785861
4          Emilia-Romagna  2174599  2292519   4467118
5   Friuli-Venezia Giulia   589215   622142   1211357
6                   Lazio  2832324  3033220   5865544
7                 Liguria   739021   804106   1543127
8               Lombardia  4949770  5154199  10103969
9                  Marche   738259   780141   1518400
10                 Molise   148764   153501    302265
11               Piemonte  2109314  2232061   4341375
12                 Puglia  1950256  2058040   4008296
13               Sardegna   800902   829572   1630474
14                Sicilia  2416233  2552177   4968410
15                Toscana  1798279  1924450   3722729
16     TrentinoAlto Adige   529824   544995   1074819
17                 Umbria   424592   455693    880285
18          Valle d'Aosta    61325    64176    125501
19                 Veneto  2403360  2504344   4907704

Correggo l'errore sul nome della regione ("TrentinoAlto Adige")

In [14]:
istat.iloc[16,0] = "Trentino-Alto Adige"
istat.iloc[16,0]

'Trentino-Alto Adige'

+ **Scaricare la tabella delle regioni italiane e visualizzarle**

In [16]:
regioni = pd.read_csv('../data/italy_regions.csv', sep=';', usecols=['regione','superficie', 'num_residenti','num_comuni', 'num_provincie'])
regioni

regione  superficie  num_residenti  num_comuni  \
0                 Abruzzo    10831.84      1326513.0       305.0   
1              Basilicata    10073.32       573694.0       131.0   
2                Calabria    15221.90      1970521.0       409.0   
3                Campania    13670.95      5850850.0       550.0   
4          Emilia-Romagna    22452.78      4448146.0       333.0   
5   Friuli-Venezia Giulia     7862.30      1221218.0       216.0   
6                   Lazio    17232.29      5888472.0       378.0   
7                 Liguria     5416.21      1571053.0       235.0   
8               Lombardia    23863.65     10008349.0      1524.0   
9                  Marche     9401.38      1543752.0       229.0   
10                 Molise     4460.65       312027.0       136.0   
11               Piemonte    25387.07      4404246.0      1201.0   
12                 Puglia    19540.90      4077166.0       258.0   
13               Sardegna    24100.02      1658138.0       377.0   
14                Sicilia    25832.39      5074261.0       390.0   
15                Toscana    22987.04      3744398.0       276.0   
16    Trentino-Alto Adige    13605.50      1059114.0       293.0   
17                 Umbria     8464.33       891181.0        92.0   
18          Valle d'Aosta     3260.90       127329.0        74.0   
19                 Veneto    18407.42      4915123.0       576.0   
20                    NaN         NaN            NaN         NaN   

    num_provincie  
0             4.0  
1             2.0  
2             5.0  
3             5.0  
4             9.0  
5             4.0  
6             5.0  
7             4.0  
8            12.0  
9             5.0  
10            2.0  
11            8.0  
12            6.0  
13            5.0  
14            9.0  
15           10.0  
16            2.0  
17            2.0  
18            1.0  
19            7.0  
20            NaN

## ANALISI VALORI NULLI 
***

+ **TRATTARE I VALORI NULLI - ISTAT** A prima vista non appaiono valori nulli, meglio analizzarli nello specifico.

In [17]:
istat.isnull().any() #non ci sono valori nulli in nessuna riga

Territorio    False
Maschi        False
Femmine       False
Totale        False
dtype: bool

+ **TRATTARE I VALORI NULLI - REGIONI** Come si evince dai risultati sono presenti valori mancanti, in questo caso NaN, not a number.

In [18]:
regioni.isnull().any() #in tutte le righe sono presenti valori nulli

regione          True
superficie       True
num_residenti    True
num_comuni       True
num_provincie    True
dtype: bool

In [19]:
regioni['regione'].isnull() #tutte false, tanne che per la 21esima

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20     True
Name: regione, dtype: bool

#### Il valore mancante appare nella 21° regione, che non esiste. Perciò possiamo senza problemi eliminare l'ultima riga contenente valori NaN.

In [20]:
regioni.dropna(inplace=True) #dataset finale senza NaN
regioni

regione  superficie  num_residenti  num_comuni  \
0                 Abruzzo    10831.84      1326513.0       305.0   
1              Basilicata    10073.32       573694.0       131.0   
2                Calabria    15221.90      1970521.0       409.0   
3                Campania    13670.95      5850850.0       550.0   
4          Emilia-Romagna    22452.78      4448146.0       333.0   
5   Friuli-Venezia Giulia     7862.30      1221218.0       216.0   
6                   Lazio    17232.29      5888472.0       378.0   
7                 Liguria     5416.21      1571053.0       235.0   
8               Lombardia    23863.65     10008349.0      1524.0   
9                  Marche     9401.38      1543752.0       229.0   
10                 Molise     4460.65       312027.0       136.0   
11               Piemonte    25387.07      4404246.0      1201.0   
12                 Puglia    19540.90      4077166.0       258.0   
13               Sardegna    24100.02      1658138.0       377.0   
14                Sicilia    25832.39      5074261.0       390.0   
15                Toscana    22987.04      3744398.0       276.0   
16    Trentino-Alto Adige    13605.50      1059114.0       293.0   
17                 Umbria     8464.33       891181.0        92.0   
18          Valle d'Aosta     3260.90       127329.0        74.0   
19                 Veneto    18407.42      4915123.0       576.0   

    num_provincie  
0             4.0  
1             2.0  
2             5.0  
3             5.0  
4             9.0  
5             4.0  
6             5.0  
7             4.0  
8            12.0  
9             5.0  
10            2.0  
11            8.0  
12            6.0  
13            5.0  
14            9.0  
15           10.0  
16            2.0  
17            2.0  
18            1.0  
19            7.0

## ANALISI
***

#### Lo scopo dell'analisi è quello di visualizzare la densità di popolazione per regione, prima nel dataset regioni poi nel dataset unito con ISTAT per vedere se il risultato è il medesimo. Poi si visualizzano le due regioni risultati più densamente popolate in un modo coerente e ordinato.

#### PUNTO PRIMO: DENSITA' ABITATIVA
* * *

+ **Visualizzare la densità di popolazione nel dataset regione**

In [21]:
regioni = regioni.set_index('regione')
densita_pop = regioni['num_residenti']/regioni['superficie']
densita_pop

regione
Abruzzo                  122.464235
Basilicata                56.951829
Calabria                 129.453025
Campania                 427.976841
Emilia-Romagna           198.111147
Friuli-Venezia Giulia    155.325795
Lazio                    341.711519
Liguria                  290.065009
Lombardia                419.397242
Marche                   164.204830
Molise                    69.951016
Piemonte                 173.483825
Puglia                   208.647810
Sardegna                  68.802350
Sicilia                  196.430179
Toscana                  162.891699
Trentino-Alto Adige       77.844548
Umbria                   105.286656
Valle d'Aosta             39.047196
Veneto                   267.018572
dtype: float64

#### Le regioni con amggiore densità abitativa sono la Campania (427,97) e la Lombardia (419,39), che sono anche due delle regioni più popolose d'Italia.

+ **Unire i due dataset**

In [22]:
regioni_pop=pd.merge(istat, regioni, left_on="Territorio", right_on="regione").drop('num_residenti', axis=1)
regioni_pop

Territorio   Maschi  Femmine    Totale  superficie  num_comuni  \
0                 Abruzzo   637388   668382   1305770    10831.84       305.0   
1              Basilicata   273941   282993    556934    10073.32       131.0   
2                Calabria   942644   982057   1924701    15221.90       409.0   
3                Campania  2820555  2965306   5785861    13670.95       550.0   
4          Emilia-Romagna  2174599  2292519   4467118    22452.78       333.0   
5   Friuli-Venezia Giulia   589215   622142   1211357     7862.30       216.0   
6                   Lazio  2832324  3033220   5865544    17232.29       378.0   
7                 Liguria   739021   804106   1543127     5416.21       235.0   
8               Lombardia  4949770  5154199  10103969    23863.65      1524.0   
9                  Marche   738259   780141   1518400     9401.38       229.0   
10                 Molise   148764   153501    302265     4460.65       136.0   
11               Piemonte  2109314  2232061   4341375    25387.07      1201.0   
12                 Puglia  1950256  2058040   4008296    19540.90       258.0   
13               Sardegna   800902   829572   1630474    24100.02       377.0   
14                Sicilia  2416233  2552177   4968410    25832.39       390.0   
15                Toscana  1798279  1924450   3722729    22987.04       276.0   
16    Trentino-Alto Adige   529824   544995   1074819    13605.50       293.0   
17                 Umbria   424592   455693    880285     8464.33        92.0   
18          Valle d'Aosta    61325    64176    125501     3260.90        74.0   
19                 Veneto  2403360  2504344   4907704    18407.42       576.0   

    num_provincie  
0             4.0  
1             2.0  
2             5.0  
3             5.0  
4             9.0  
5             4.0  
6             5.0  
7             4.0  
8            12.0  
9             5.0  
10            2.0  
11            8.0  
12            6.0  
13            5.0  
14            9.0  
15           10.0  
16            2.0  
17            2.0  
18            1.0  
19            7.0

+ **Utilizzare funzione melt per creare colonna Sesso e Popolazione**

In [23]:
regioni_pop_tot=regioni_pop.melt(
               id_vars=["Territorio", "superficie","num_comuni", "num_provincie"],  
               var_name="Sesso",             
               value_name="Popolazione")
regioni_pop_tot

Territorio  superficie  num_comuni  num_provincie    Sesso  \
0                 Abruzzo    10831.84       305.0            4.0   Maschi   
1              Basilicata    10073.32       131.0            2.0   Maschi   
2                Calabria    15221.90       409.0            5.0   Maschi   
3                Campania    13670.95       550.0            5.0   Maschi   
4          Emilia-Romagna    22452.78       333.0            9.0   Maschi   
5   Friuli-Venezia Giulia     7862.30       216.0            4.0   Maschi   
6                   Lazio    17232.29       378.0            5.0   Maschi   
7                 Liguria     5416.21       235.0            4.0   Maschi   
8               Lombardia    23863.65      1524.0           12.0   Maschi   
9                  Marche     9401.38       229.0            5.0   Maschi   
10                 Molise     4460.65       136.0            2.0   Maschi   
11               Piemonte    25387.07      1201.0            8.0   Maschi   
12                 Puglia    19540.90       258.0            6.0   Maschi   
13               Sardegna    24100.02       377.0            5.0   Maschi   
14                Sicilia    25832.39       390.0            9.0   Maschi   
15                Toscana    22987.04       276.0           10.0   Maschi   
16    Trentino-Alto Adige    13605.50       293.0            2.0   Maschi   
17                 Umbria     8464.33        92.0            2.0   Maschi   
18          Valle d'Aosta     3260.90        74.0            1.0   Maschi   
19                 Veneto    18407.42       576.0            7.0   Maschi   
20                Abruzzo    10831.84       305.0            4.0  Femmine   
21             Basilicata    10073.32       131.0            2.0  Femmine   
22               Calabria    15221.90       409.0            5.0  Femmine   
23               Campania    13670.95       550.0            5.0  Femmine   
24         Emilia-Romagna    22452.78       333.0            9.0  Femmine   
25  Friuli-Venezia Giulia     7862.30       216.0            4.0  Femmine   
26                  Lazio    17232.29       378.0            5.0  Femmine   
27                Liguria     5416.21       235.0            4.0  Femmine   
28              Lombardia    23863.65      1524.0           12.0  Femmine   
29                 Marche     9401.38       229.0            5.0  Femmine   
30                 Molise     4460.65       136.0            2.0  Femmine   
31               Piemonte    25387.07      1201.0            8.0  Femmine   
32                 Puglia    19540.90       258.0            6.0  Femmine   
33               Sardegna    24100.02       377.0            5.0  Femmine   
34                Sicilia    25832.39       390.0            9.0  Femmine   
35                Toscana    22987.04       276.0           10.0  Femmine   
36    Trentino-Alto Adige    13605.50       293.0            2.0  Femmine   
37                 Umbria     8464.33        92.0            2.0  Femmine   
38          Valle d'Aosta     3260.90        74.0            1.0  Femmine   
39                 Veneto    18407.42       576.0            7.0  Femmine   
40                Abruzzo    10831.84       305.0            4.0   Totale   
41             Basilicata    10073.32       131.0            2.0   Totale   
42               Calabria    15221.90       409.0            5.0   Totale   
43               Campania    13670.95       550.0            5.0   Totale   
44         Emilia-Romagna    22452.78       333.0            9.0   Totale   
45  Friuli-Venezia Giulia     7862.30       216.0            4.0   Totale   
46                  Lazio    17232.29       378.0            5.0   Totale   
47                Liguria     5416.21       235.0            4.0   Totale   
48              Lombardia    23863.65      1524.0           12.0   Totale   
49                 Marche     9401.38       229.0            5.0   Totale   
50                 Molise     4460.65       136.0            2.0   Totale   
51         

+ **Indicizzo secondo territorio**

In [24]:
regioni_pop_tot1 = regioni_pop_tot.set_index(['Territorio'])
regioni_pop_tot1

superficie  num_comuni  num_provincie    Sesso  \
Territorio                                                              
Abruzzo                  10831.84       305.0            4.0   Maschi   
Basilicata               10073.32       131.0            2.0   Maschi   
Calabria                 15221.90       409.0            5.0   Maschi   
Campania                 13670.95       550.0            5.0   Maschi   
Emilia-Romagna           22452.78       333.0            9.0   Maschi   
Friuli-Venezia Giulia     7862.30       216.0            4.0   Maschi   
Lazio                    17232.29       378.0            5.0   Maschi   
Liguria                   5416.21       235.0            4.0   Maschi   
Lombardia                23863.65      1524.0           12.0   Maschi   
Marche                    9401.38       229.0            5.0   Maschi   
Molise                    4460.65       136.0            2.0   Maschi   
Piemonte                 25387.07      1201.0            8.0   Maschi   
Puglia                   19540.90       258.0            6.0   Maschi   
Sardegna                 24100.02       377.0            5.0   Maschi   
Sicilia                  25832.39       390.0            9.0   Maschi   
Toscana                  22987.04       276.0           10.0   Maschi   
Trentino-Alto Adige      13605.50       293.0            2.0   Maschi   
Umbria                    8464.33        92.0            2.0   Maschi   
Valle d'Aosta             3260.90        74.0            1.0   Maschi   
Veneto                   18407.42       576.0            7.0   Maschi   
Abruzzo                  10831.84       305.0            4.0  Femmine   
Basilicata               10073.32       131.0            2.0  Femmine   
Calabria                 15221.90       409.0            5.0  Femmine   
Campania                 13670.95       550.0            5.0  Femmine   
Emilia-Romagna           22452.78       333.0            9.0  Femmine   
Friuli-Venezia Giulia     7862.30       216.0            4.0  Femmine   
Lazio                    17232.29       378.0            5.0  Femmine   
Liguria                   5416.21       235.0            4.0  Femmine   
Lombardia                23863.65      1524.0           12.0  Femmine   
Marche                    9401.38       229.0            5.0  Femmine   
Molise                    4460.65       136.0            2.0  Femmine   
Piemonte                 25387.07      1201.0            8.0  Femmine   
Puglia                   19540.90       258.0            6.0  Femmine   
Sardegna                 24100.02       377.0            5.0  Femmine   
Sicilia                  25832.39       390.0            9.0  Femmine   
Toscana                  22987.04       276.0           10.0  Femmine   
Trentino-Alto Adige      13605.50       293.0            2.0  Femmine   
Umbria                    8464.33        92.0            2.0  Femmine   
Valle d'Aosta             3260.90        74.0            1.0  Femmine   
Veneto                   18407.42       576.0            7.0  Femmine   
Abruzzo                  10831.84       305.0            4.0   Totale   
Basilicata               10073.32       131.0            2.0   Totale   
Calabria                 15221.90       409.0            5.0   Totale   
Campania                 13670.95       550.0            5.0   Totale   
Emilia-Romagna           22452.78       333.0            9.0   Totale   
Friuli-Venezia Giulia     7862.30       216.0            4.0   Totale   
Lazio                    17232.29       378.0            5.0   Totale   
Liguria                   5416.21       235.0            4.0   Totale   
Lombardia                23863.65      1524.0           12.0   Totale   
Marche                    9401.38       229.0            5.0   Totale   
Molise                    4460.65       136.0            2.0   Totale   
Piemonte                 25387.07      1201.0            8.0   Totale   
Puglia                   19540.90       258.0            6.0   Totale   
Sardegna

+ **Selezionare solo la popolazione totale, non suddivisa tra maschi e femmine**

In [12]:
regioni_pop_tot1 = regioni_pop_tot1.loc[regioni_pop_tot1['Sesso']=='Totale']
regioni_pop_tot1

superficie  num_comuni  num_provincie   Sesso  \
Territorio                                                             
Abruzzo                  10831.84       305.0            4.0  Totale   
Basilicata               10073.32       131.0            2.0  Totale   
Calabria                 15221.90       409.0            5.0  Totale   
Campania                 13670.95       550.0            5.0  Totale   
Emilia-Romagna           22452.78       333.0            9.0  Totale   
Friuli-Venezia Giulia     7862.30       216.0            4.0  Totale   
Lazio                    17232.29       378.0            5.0  Totale   
Liguria                   5416.21       235.0            4.0  Totale   
Lombardia                23863.65      1524.0           12.0  Totale   
Marche                    9401.38       229.0            5.0  Totale   
Molise                    4460.65       136.0            2.0  Totale   
Piemonte                 25387.07      1201.0            8.0  Totale   
Puglia                   19540.90       258.0            6.0  Totale   
Sardegna                 24100.02       377.0            5.0  Totale   
Sicilia                  25832.39       390.0            9.0  Totale   
Toscana                  22987.04       276.0           10.0  Totale   
Umbria                    8464.33        92.0            2.0  Totale   
Valle d'Aosta             3260.90        74.0            1.0  Totale   
Veneto                   18407.42       576.0            7.0  Totale   

                       Popolazione  
Territorio                          
Abruzzo                    1305770  
Basilicata                  556934  
Calabria                   1924701  
Campania                   5785861  
Emilia-Romagna             4467118  
Friuli-Venezia Giulia      1211357  
Lazio                      5865544  
Liguria                    1543127  
Lombardia                 10103969  
Marche                     1518400  
Molise                      302265  
Piemonte                   4341375  
Puglia                     4008296  
Sardegna                   1630474  
Sicilia                    4968410  
Toscana                    3722729  
Umbria                      880285  
Valle d'Aosta               125501  
Veneto                     4907704

+ **Visualizzare la densità abitativa**

In [13]:
densita_pop2 = regioni_pop_tot1['Popolazione']/regioni_pop_tot1['superficie']
densita_pop2

Territorio
Abruzzo                  120.549233
Basilicata                55.288028
Calabria                 126.442888
Campania                 423.223039
Emilia-Romagna           198.956120
Friuli-Venezia Giulia    154.071582
Lazio                    340.380994
Liguria                  284.909005
Lombardia                423.404173
Marche                   161.508204
Molise                    67.762546
Piemonte                 171.007328
Puglia                   205.123408
Sardegna                  67.654467
Sicilia                  192.332572
Toscana                  161.949037
Umbria                   103.999371
Valle d'Aosta             38.486614
Veneto                   266.615528
dtype: float64

#### Le regioni con la maggiore densità abitativa sono sempre la Campania (423,22) e la Lombardia (423,40), questa volta con valori differenti. Non essendo lo scarto troppo significativo, il motivo può essere un piccolo errore di conteggio (fermo restando che i dati ISTAT sono i più affidabili).

#### PUNTO SECONDO: VISUALIZZARE LE DUE REGIONI PIU' DENSAMENTE POPOLATE
* * *

+ **Selezionare le due regioni più densamente popolate**

In [14]:
camp_lomb = regioni_pop_tot.loc[(regioni_pop_tot['Territorio']== 'Campania') | (regioni_pop_tot['Territorio']== 'Lombardia')]
camp_lomb

Territorio  superficie  num_comuni  num_provincie    Sesso  Popolazione
3    Campania    13670.95       550.0            5.0   Maschi      2820555
8   Lombardia    23863.65      1524.0           12.0   Maschi      4949770
22   Campania    13670.95       550.0            5.0  Femmine      2965306
27  Lombardia    23863.65      1524.0           12.0  Femmine      5154199
41   Campania    13670.95       550.0            5.0   Totale      5785861
46  Lombardia    23863.65      1524.0           12.0   Totale     10103969

+ **Visuliazziamolo un po' meglio**

In [15]:
camp_lomb_fin = camp_lomb.groupby(['Sesso', 'Territorio'])

In [16]:
camp_lomb_fin.sum()

superficie  num_comuni  num_provincie  Popolazione
Sesso   Territorio                                                    
Femmine Campania      13670.95       550.0            5.0      2965306
        Lombardia     23863.65      1524.0           12.0      5154199
Maschi  Campania      13670.95       550.0            5.0      2820555
        Lombardia     23863.65      1524.0           12.0      4949770
Totale  Campania      13670.95       550.0            5.0      5785861
        Lombardia     23863.65      1524.0           12.0     10103969

+ **In alternativa**

In [17]:
camp_lomb1 = regioni_pop_tot.query('Territorio=="Campania" | Territorio=="Lombardia"')
camp_lomb1

Territorio  superficie  num_comuni  num_provincie    Sesso  Popolazione
3    Campania    13670.95       550.0            5.0   Maschi      2820555
8   Lombardia    23863.65      1524.0           12.0   Maschi      4949770
22   Campania    13670.95       550.0            5.0  Femmine      2965306
27  Lombardia    23863.65      1524.0           12.0  Femmine      5154199
41   Campania    13670.95       550.0            5.0   Totale      5785861
46  Lombardia    23863.65      1524.0           12.0   Totale     10103969

In [18]:
camp_lomb2 = camp_lomb1.groupby(['Territorio', 'Sesso'])

In [19]:
camp_lomb2.sum()

superficie  num_comuni  num_provincie  Popolazione
Territorio Sesso                                                      
Campania   Femmine    13670.95       550.0            5.0      2965306
           Maschi     13670.95       550.0            5.0      2820555
           Totale     13670.95       550.0            5.0      5785861
Lombardia  Femmine    23863.65      1524.0           12.0      5154199
           Maschi     23863.65      1524.0           12.0      4949770
           Totale     23863.65      1524.0           12.0     10103969

#### COMMENTO: La regione Campania presenta 5785861 abitanti, con una sovrarappresentazione femminile rispetto che maschile. Ha 5 province (Napoli, Avellino, Salerno, Caserta, Benevento) e 550 comuni. Conta una superficie di 13670.95 km quadrati. La regione Lombardia presenta 10103969 abitanti, con una sovrarappresentazione femminile rispetto che maschile (anche in questo caso, il motivo può essere collegato alla maggiore longevità). Ha 12 province (Milano, Bergamo, Monza Brianza, Lodi, Cremona, Sondrio, Mantova, Pavia, Brescia, Lecco, Como, Varese) e 1524 comuni. Conta una superficie maggiore della Campania, di 23863.65 km quadrati (quasi il doppio).